infer dynamic shapes through additional inputs

avikchaudhuri opened the pull request:
Summary:
Instead of explicitly specifying dynamic shapes, it is possible to infer them from additional example inputs. Together with the example inputs provided to export, we can basically make any varying dim dynamic and keep any fixed dim static. This should be useful for prod scenarios that have access to tests and/or profiling data, yet are somewhat removed from the model authoring process.

However this alone is not satisfactory: the exported program by design has only one graph, representing one path through the model, and we cannot necessarily guarantee that this graph works for the additional example inputs because different guards might have been created if we had exported with them instead (corresponding to different traced paths). However, checking that the additional example inputs satisfy the guards created by the original export should be sufficient for generalization.

Now, while we don't preserve all guards in the exported program, we do check a subset of them as part of input matching. So we add a verification step at the end of export when such additional example inputs are provided. This should be enough for now.

Test Plan: added test (positive and negative cases)

Differential Revision: D72001771

bobrenjc93 commented:
I wrote this in the DS chat, just pasting here in case you missed it:

I actually asked @pianpwk to help me take this to export a few days ago lol. I'm indifferent to what approach we take, but it'd be nice to consolidate on a shared library for inferring dynamism. https://github.com/pytorch/pytorch/pull/147981 is the dynamism library we use for the various stances in compile (`aot_eager_then_compile` and `eager_then_compile`)

avikchaudhuri commented:
> I wrote this in the DS chat, just pasting here in case you missed it:
> 
> I actually asked @pianpwk to help me take this to export a few days ago lol. I'm indifferent to what approach we take, but it'd be nice to consolidate on a shared library for inferring dynamism. #147981 is the dynamism library we use for the various stances in compile (`aot_eager_then_compile` and `eager_then_compile`)

Ack. It's not immediately clear how to reuse that library here. I'm following the pattern of `ShapesCollection` here, desugaring to `Dim`-based dynamic shapes. The actual logic of multiple specs for the same key is then handled by a simple tree map; not worried about duplication here, it's small and trivial. The big difference in this approach and that library is that what data structure to desugar to. Unification of that base representation would be an orthogonal project, basically needed some refactoring on both sides. Not planning to do that here FYI.

@pianpwk Does that make sense?

bobrenjc93 commented on torch/export/dynamic_shapes.py:
```diff
@@ -702,6 +703,84 @@ def find_shape(path, t):
+        assert (
+            kwargs is None or type(kwargs) is dict
+        ), f"Representative kwargs {kwargs} must be None or a dict"
+        self._examples.append((args, kwargs))
```
should we fakify these example inputs? otherwise we may unintentionally bloat memory?

avikchaudhuri committed:
infer dynamic shapes through additional inputs (#150144)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/150144

Instead of explicitly specifying dynamic shapes, it is possible to infer them from additional example inputs. Together with the example inputs provided to export, we can basically make any varying dim dynamic and keep any fixed dim static. This should be useful for prod scenarios that have access to tests and/or profiling data, yet are somewhat removed from the model authoring process.

However this alone is not satisfactory: the exported program by design has only one graph, representing one path through the model, and we cannot necessarily guarantee that this graph works for the additional example inputs because different guards might have been created if we had exported with them instead (corresponding to different traced paths). However, checking that the additional example inputs satisfy the guards created by the original export should be sufficient for generalization.

Now, while we don't preserve all guards in the exported program, we do check a subset of them as part of input matching. So we add a verification step at the end of export when such additional example inputs are provided. This should be enough for now.

Test Plan: added test (positive and negative cases)

Reviewed By: bobrenjc93

Differential Revision: D72001771
